variables.tf

```diff
@@ -14,104 +14,120 @@
  * limitations under the License.
  */
 
-variable project {
+variable "project" {
   description = "The project to deploy to, if not set the default provider project is used."
-  default     = ""
+  type        = string
 }
 
-variable region {
+variable "region" {
   description = "Region for cloud resources"
+  type        = string
   default     = "us-central1"
 }
 
-variable ip_version {
+variable "ip_version" {
   description = "IP version for the Global address (IPv4 or v6) - Empty defaults to IPV4"
+  type        = string
   default     = ""
 }
 
-variable firewall_networks {
+variable "firewall_networks" {
   description = "Names of the networks to create firewall rules in"
-  type        = "list"
-  default     = ["default"]
+  type        = list(string)
+  default     = [
+    "default"]
 }
 
-variable firewall_projects {
+variable "firewall_projects" {
   description = "Names of the projects to create firewall rules in"
-  type        = "list"
-  default     = ["default"]
+  type        = list(string)
+  default     = [
+    "default"]
 }
 
-variable name {
+variable "name" {
   description = "Name for the forwarding rule and prefix for supporting resources"
+  type        = string
 }
 
-variable target_tags {
+variable "target_tags" {
   description = "List of target tags for health check firewall rule."
-  type        = "list"
+  type        = list(string)
 }
 
-variable backends {
+variable "backends" {
   description = "Map backend indices to list of backend maps."
-  type        = "map"
+  type        = map(list(object({
+    group = string
```

Comment:
Let's add all of the supported keys here.

Reply:
@aaron-lane this would brake all the examples down, should we modify them ? How do you feel about my suggestion in the [other conversation](https://github.com/terraform-google-modules/terraform-google-lb-http/pull/52#discussion_r329080136) for the same?

Reply:
Let me clarify this request. The internal object should include [each key which is retrieved in the configuration](https://github.com/terraform-google-modules/terraform-google-lb-http/pull/52/files#diff-7a370d8342e7203b805911c92454f0f4R92-R100), and each key should be typed correctly. Only specifying `group` in this signature is not valid because any other key in an object passed to this variable will be ignored.

```hcl
    balancing_mode = string
    capacity_scaler = string
    description = string
    group = string
    max_connections = number
    max_connections_per_instance = number
    max_rate = number
    max_rate_per_instance = number
    max_utilization = number
```

Calling configurations can provide objects with some values to `null` if they only need to configure some keys. Additionally, the code in `main.tf` which uses this object will not need to provide `null` values because that will be handled by the calling configuration.

Examples should be updated to adhere to the new object signature.

Reply:
@aaron-lane, thank you for clarification! 
Enforced backend params validation, removed default `null` from main, adjusted examples. [Fixed](https://github.com/terraform-google-modules/terraform-google-lb-http/compare/42602b96de3e1eaa0f974e3f613a228edf64d322..c9213ad6284c1e30e30bd142519f80ab3f81e76c)